src/protocols/tacacs/encode.c

```diff
@@ -39,83 +40,56 @@
 /**
  *	Encode a TACACS+ 'arg_N' fields.
  */
-static uint8_t tacacs_encode_body_arg_n_len(fr_pair_t *vps, fr_dict_attr_t const *da, uint8_t **body, uint8_t const *end)
+static uint8_t tacacs_encode_body_arg_n_len(fr_dbuff_t *dbuff, fr_pair_t *vps, fr_dict_attr_t const *da)
 {
-	uint8_t *p = *body;
-	uint8_t arg_cnt = 0;
-	size_t total = 0;
-	fr_pair_t *vp;
+	uint8_t     arg_cnt = 0;
+	fr_pair_t   *vp;
 	fr_cursor_t cursor;
 
-	fr_assert(body != NULL);
-
 	for (vp = fr_cursor_init(&cursor, &vps);
 	     vp;
 	     vp = fr_cursor_next(&cursor)) {
 		if (arg_cnt == 255) break;
 
-		if (vp->da != da) continue;
-
-		if (vp->vp_length > 0xff) continue;
-
-		if ((p + 1 + vp->vp_length + total) > end) break;
+		if (vp->da != da || vp->vp_length > 0xff) continue;
 
 		/* Append the <arg_N_len> fields length */
-		*p++ = vp->vp_length;
-		total += vp->vp_length;
+		fr_dbuff_in(dbuff, (uint8_t) vp->vp_length);
+
 		arg_cnt++;
 	}
 
-	*body = p;
-
 	return arg_cnt;
 }
 
-static void tacacs_encode_body_arg_n(fr_pair_t *vps, fr_dict_attr_t const *da, uint8_t **body, uint8_t const *end)
+static ssize_t tacacs_encode_body_arg_n(fr_dbuff_t *dbuff, fr_pair_t *vps, fr_dict_attr_t const *da)
 {
-	uint8_t *p = *body;
-	uint8_t arg_cnt = 0;
-	fr_pair_t *vp;
+	fr_pair_t   *vp;
 	fr_cursor_t cursor;
 
-	fr_assert(body != NULL);
-
 	for (vp = fr_cursor_init(&cursor, &vps);
 	     vp;
 	     vp = fr_cursor_next(&cursor)) {
-		if (arg_cnt == 255) break;
-
-		if (vp->da != da) continue;
-
-		if (vp->vp_length > 0xff) continue;
-
-		if ((p + vp->vp_length) > end) break;
+		if (vp->da != da || vp->vp_length > 0xff) continue;
 
 		/* Append the <arg_N> field */
-		memcpy(p, vp->vp_strvalue, vp->vp_length);
-		p += vp->vp_length;
+		FR_DBUFF_MEMCPY_IN_RETURN(dbuff, vp->vp_strvalue, vp->vp_length);
 	}
 
-	*body = p;
+	return 0;
 }
 
 /*
  *	Encode a TACACS+ field.
  */
-static ssize_t tacacs_encode_field(fr_pair_t *vps, fr_dict_attr_t const *da, uint8_t **data, uint8_t const *end, size_t max_len)
+static ssize_t tacacs_encode_field(fr_dbuff_t *dbuff, fr_pair_t *vps, fr_dict_attr_t const *da, size_t max_len)
 {
-	fr_pair_t *vp;
-	uint8_t *p = *data;
+	fr_pair_t 		*vp;
 
 	vp = fr_pair_find_by_da(vps, da);
-	if (!vp || !vp->vp_length) return 0;
-
-	if (vp->vp_length > max_len) return 0;
-	if ((p + vp->vp_length) > end) return 0;
+	if (!vp || !vp->vp_length || (vp->vp_length > max_len)) return 0;
```

Comment:
brackets... and I think max_len can be done with FR_DBUFF_MAX now?

Reply:
We need that `max_len` check because we have some fields with a length of `0xff` and others `0xffff`